### PR TITLE
Fix TPL search order and error reporting

### PR DESCRIPTION
**Pull request by jjwilke**
There are subtle issues due to missing `NO_DEFAULT_PATH` that lead to system libraries being preferred over explicitly given paths. There were also subtle issues in `FIND_PACKAGE_HANDLE_...` error reporting.

### CODE REVIEW

**Comment by jjwilke** [comment]
I believe this was also an issue before the last TPL Pull Request. This is not a new bug.

**Comment by masterleinad** [comment]
```
--  VOLTA70
-- Could NOT find CUDA (missing: CUDA_FOUND_LIBRARIES) 
CMake Error at /opt/cmake/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find TPLCUDA (missing: TPL_CUDA_VALID)
Call Stack (most recent call first):
  /opt/cmake/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/kokkos_functions.cmake:536 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  cmake/Modules/FindTPLCUDA.cmake:3 (KOKKOS_FIND_IMPORTED)
  cmake/kokkos_functions.cmake:189 (FIND_PACKAGE)
  cmake/kokkos_tpls.cmake:35 (KOKKOS_IMPORT_TPL)
  cmake/kokkos_tribits.cmake:178 (INCLUDE)
  CMakeLists.txt:133 (KOKKOS_SETUP_BUILD_ENVIRONMENT)

```

**Comment by jjwilke** [comment]
If it's not already doing so (I couldnt find it), it would be helpful for the Jenkins/Travis to just print the full environment for debugging purposes. Anyhow, I believe this is fixed?

**Comment by dalg24** [comment]
> If it's not already doing so (I couldnt find it), it would be helpful for the Jenkins/Travis to just print the full environment for debugging purposes.

You have full control about what is being ran in the CI.  You can add a `sh` step that does so as needed.
Your other option if you have docker on your machine is
```
cd scripts/docker
docker build -t <give_your_image_a_name> -f <dockerfile_as_in_the_jenkinsfile> --build-args <build_arguments_as_in_the_jenkinsfile> .
docker run --rm <give_your_image_a_name> env
```
PS: you do not need CUDA on your machine to test.

File: cmake/kokkos_functions.cmake [hunk 1]
```diff
@@ -314,6 +314,151 @@ MACRO(kokkos_create_imported_tpl NAME)
...
+    FIND_PATH(${VAR_NAME} ${HEADER} PATHS ${KOKKOS_${TPL_NAME}_DIR}/include NO_DEFAULT_PATH)
+    SET(HAVE_CUSTOM_PATHS TRUE)
+  ENDIF()
+
+  IF (NOT ${HEADER}_FOUND AND TPL_PATHS)
+    #we got custom paths
+    #ONLY look in these paths and nowhere else
+    FIND_PATH(${VAR_NAME} ${HEADER} PATHS ${TPL_PATHS} NO_DEFAULT_PATH)
+    SET(HAVE_CUSTOM_PATHS TRUE)
+  ENDIF()
```
Comment (dalg24) [inline cmake/kokkos_functions.cmake]: Implicitly `TPL_PATHS` > `KOKKOS_${TPL_NAME}_DIR` >`${TPL_NAME}_ROOT` (where > indicates that a variable takes precedence over another)

File: cmake/kokkos_functions.cmake [hunk 2]
```diff
@@ -364,11 +522,21 @@ ENDMACRO()
...
    ${ARGN})
 
+  IF(NOT TPL_MODULE_NAME)
+    SET(TPL_MODULE_NAME TPL${NAME})
+  ENDIF()
+
+  IF (TPL_ALLOW_SYSTEM_PATH_FALLBACK)
+    SET(ALLOW_PATH_FALLBACK_OPT ALLOW_SYSTEM_PATH_FALLBACK)
+  ELSE()
+    SET(ALLOW_PATH_FALLBACK_OPT)
```
Comment (dalg24) [inline cmake/kokkos_functions.cmake]: I suppose you don't even need to define it in that case.

File: cmake/kokkos_functions.cmake [hunk 3]
```diff
@@ -314,6 +314,151 @@ MACRO(kokkos_create_imported_tpl NAME)
...
+# or default system paths depending on options. In constrast to CMake
+# default, custom paths are prioritized over system paths
+#
+# Usage::
+#
+#   KOKKOS_FIND_HEADER(
+#     <VAR_NAME>
+#     <HEADER>
+#     <TPL_NAME>
+#     ...
```
Comment (dalg24) [inline cmake/kokkos_functions.cmake]: You forgot to close the parentheses.

Also the signature reads like this
```
KOKKOS_FIND_HEADER(
    <VAR>
    <HEADER>
    <TPL>
    [ALLOW_SYSTEM_PATH_FALLBACK]
    [PATHS path1 [path2 ...]]
)
```

File: cmake/kokkos_functions.cmake [hunk 4]
```diff
@@ -379,70 +547,73 @@ MACRO(kokkos_find_imported NAME)
...
+      SET(${NAME}_FOUND_LIBRARIES ${${LIB}_LOCATION}) 
       BREAK()
     ENDIF()
   ENDFOREACH()
 
   INCLUDE(FindPackageHandleStandardArgs)
+  # I don't know a good way to chain all of the relevant variables into
+  # a single call to find_package_handle. find_package only cares
+  # about the failure of the LAST call to find_package_handle_...
+  # so if there as an error I need to skip the rest
```
Comment (dalg24) [inline cmake/kokkos_functions.cmake]: Look at the 2nd signature in the [doc](https://cmake.org/cmake/help/latest/module/FindPackageHandleStandardArgs.html?highlight=find_package_handle_standard_args#command:find_package_handle_standard_args), what you want is `REQUIRED_VARS <required-var>...`

File: cmake/kokkos_functions.cmake (hunk 1 above)
Comment (jjwilke) [reply cmake/kokkos_functions.cmake]: I set this as `ROOT > DIR > PATHS > SYSTEM`. I added some extra documentation for that. This seemed the most logical way of doing things?  This basically means user-provided > kokkos guesses > system defaults.  I can reorder, if anyone feels strongly this is not the right search order.

File: cmake/kokkos_functions.cmake (hunk 4 above)
Comment (jjwilke) [reply cmake/kokkos_functions.cmake]: Fixed. I love CMake argument packing.

File: cmake/kokkos_functions.cmake (hunk 3 above)
Comment (jjwilke) [reply cmake/kokkos_functions.cmake]: Done.

**Pull request merged**